File: The_Coupon_Code.js
/*
Story
Your online store likes to give out coupons for special occasions. Some customers try to cheat the system by entering invalid codes or using expired coupons.

Task
Your mission:
Write a function called checkCoupon which verifies that a coupon code is valid and not expired.

A coupon is no more valid on the day AFTER the expiration date. All dates will be passed as strings in this format: "MONTH DATE, YEAR".

Examples:
checkCoupon("123", "123", "July 9, 2015", "July 9, 2015")  ===  true
checkCoupon("123", "123", "July 9, 2015", "July 2, 2015")  ===  false
*/

function checkCoupon(enteredCode, correctCode, currentDate, expirationDate){
  monthList = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"];

  if (enteredCode == correctCode) {
    currentDate = currentDate.split(' ');
    expirationDate = expirationDate.split(' ');

    console.log(currentDate[2]);
    console.log(expirationDate[2]);
    if ((Number(currentDate[2]) <= Number(expirationDate[2]))) {
      console.log(monthList.indexOf(currentDate[0]));
      console.log(monthList.indexOf(expirationDate[0]));
      if (monthList.indexOf(currentDate[0]) < monthList.indexOf(expirationDate[0])) {
        return true;
      }
      else if (monthList.indexOf(currentDate[0]) == monthList.indexOf(expirationDate[0])) {
        if (Number(currentDate[1].replace(',', '')) <= Number(expirationDate[1].replace(',', ''))) {
          return true;
        }
      }
    }
  }
  return false;
}

// Testing
console.log(checkCoupon("123", "123", "July 9, 2015", "July 9, 2015")); //true
//console.log(checkCoupon("123", "123", "July 9, 2015", "July 2, 2015")); //false
//console.log(checkCoupon('123','123','September 5, 2014','October 1, 2014')) //true
//console.log(checkCoupon('123a','123','September 5, 2014','October 1, 2014')) //false

File: The_Coupon_Code.py
#!/usr/bin/env python3

"""
Story
Your online store likes to give out coupons for special occasions. Some customers try to cheat the system by entering invalid codes or using expired coupons.

Task
Your mission:
Write a function called checkCoupon which verifies that a coupon code is valid and not expired.

A coupon is no more valid on the day AFTER the expiration date. All dates will be passed as strings in this format: "MONTH DATE, YEAR".

Examples:
checkCoupon("123", "123", "July 9, 2015", "July 9, 2015")  == True
checkCoupon("123", "123", "July 9, 2015", "July 2, 2015")  == False
"""

def check_coupon(entered_code, correct_code, current_date, expiration_date):
    if entered_code == correct_code:
        monthList = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
        current_date = current_date.split(' ')
        expiration_date = expiration_date.split(' ')

        if int(current_date[2]) <= int(expiration_date[2]):
            if monthList.index(current_date[0]) < monthList.index(expiration_date[0]):
                return True
            if monthList.index(current_date[0]) == monthList.index(expiration_date[0]):
                if int(current_date[1][:-1]) <= int(expiration_date[1][:-1]):
                    return True
    return False

# Testing
print(check_coupon("123", "123", "July 9, 2015", "July 9, 2015")) #True
print(check_coupon("123", "123", "July 9, 2015", "July 2, 2015")) #False
print(check_coupon('123','123','September 5, 2014','October 1, 2014')) #True
print(check_coupon('123a','123','September 5, 2014','October 1, 2014')) #False
